### ISSUE TITLE
IIFE w/ throw error in default value of destructuring assignment produces false positive 'unreachable code detected.'

### ISSUE THREAD

**Issue by Huhspaniel** [microsoft/TypeScript#49759]
# Bug Report

If an IIFE which throws an error is used as the default value of a destructuring assignment, any code after the assignment is marked as unreachable. If a declared function with return type `never` is used instead, this does not occur.
### 🔎 Search Terms

- iife
- unreachable code
- destructuring
### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about destructuring, IIFE's, type `never`, and unreachable code.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220701#code/MYewdgzgLgBAJgQygmBeGMDeMAOCBOCAtgFwwDkAbggDYCuApuTAL4DcAUB6JLJhxgx5CRNDAAU4gJRoAfFhhQAFvhAB3GGAYaAovlX5x5YcRgBLCJpCw4DAGZmtccjJZTpAGg4sxiZJ25wCBAaBgA6GhAAc3ETIilODCA)

### 💻 Code

```ts
const data =  { param: 'value' };

const {
    param = (() => { throw new Error('param is not defined') })(),
} = data;

console.log(param); // triggers 'Unreachable code detected.'
```

### 🙁 Actual behavior

Produces false positive 'Unreachable code detected' on `console.log` line. If `data.param` is defined (which it is in this example), an error is never thrown and the `console.log` line will execute. If `data.param` is of type `string | undefined`, then the resulting type after the destructuring assignment is `string`, which is correct and as expected ([additional playground link](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220701#code/CYUwxgNghgTiAEYD2A7AzgF3sKGoC54BveAB1igFsB+QzGASxQHN4BfAbgCgvl0siXeMLIVK8ALzwAFNICUkgHzF4GABYwkAd3goQOgKIxNMaQHJyMKvAZpdSLKABmTEMDMK2c+QBoubSWxcKG5eVDQkCBAAOggkZmlLKjluYSA)), bar the 'Unreachable code detected' error.
### 🙂 Expected behavior

Should not show 'unreachable code detected' because the IIFE which throws an error will only execute if `data.param` is not defined.

**Comment by fatcerberus** [comment]
> Should not show 'unreachable code detected' because the IIFE which throws an error will only execute if data.param is not defined.

TIL.  I had thought the default value in a destructuring assignment was always evaluated regardless of whether it was actually used or not.
